packages/devtools_app/lib/src/provider/eval.dart

```diff
@@ -0,0 +1,222 @@
+/// A few utilities related to evaluating dart code
+library eval;
+
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:meta/meta.dart';
+import 'package:vm_service/vm_service.dart' hide Error;
+
+import '../eval_on_dart_library.dart';
+import '../globals.dart';
+import '../inspector/inspector_service.dart';
+import 'result.dart';
+
+/// An [EvalOnDartLibrary] that has access to `provider`
+///
+/// Not suitable to be used when evaluating third-party objects, as it would
+/// otherwise not be possible to read private properties.
+final evalProvider = Provider<EvalOnDartLibrary>((ref) {
+  final eval = EvalOnDartLibrary(
+    ['package:provider/src/provider.dart'],
+    serviceManager.service,
+  );
+
+  ref.onDispose(eval.dispose);
+  return eval;
+});
+
+/// An [EvalOnDartLibrary] for custom objects.
+final libraryEvalProvider =
+    AutoDisposeProviderFamily<EvalOnDartLibrary, String>((ref, libraryPath) {
+  final eval = EvalOnDartLibrary([libraryPath], serviceManager.service);
+
+  ref.onDispose(eval.dispose);
+  return eval;
+});
+
+Result<T> parseSentinel<T>(Object value) {
+  if (value is T) {
+    return Result.data(value);
+  } else if (value is Sentinel) {
+    return Result.error(SentinelException(value));
+  } else {
+    return Result.error(
+      UnsupportedError('unkown type ${value.runtimeType}'),
+      StackTrace.current,
+    );
+  }
+}
+
+extension SafeEval on EvalOnDartLibrary {
+  Future<Instance> evalInstance(
+    String expression, {
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    final ref = await safeEval(expression, isAlive: isAlive, scope: scope);
+
+    try {
+      return getInstance(ref, isAlive);
+    } catch (err) {
+      print('did throw');
+      rethrow;
+    }
+  }
+
+  static int _nextId = 0;
+
+  /// a [safeEval] variant that does not complete until the Future evaluated completes
+  ///
+  /// The expression **must** return a [Future].
+  ///
+  /// Works only if the evaluated library includes the following code:
+  ///
+  /// ```dart
+  /// Future<T> $await<T>(Future<T> future, String id) async {
+  ///   try {
+  ///     return await future;
+  ///   } finally {
+  ///     postEvent('future_completed', {'id': id});
+  ///   }
+  /// }
+  /// ```
+  Future<InstanceRef> awaitEval(
```

Comment:
Same for this nice helper. Can you also provide a link to the dartsdk bug you filed about providing a way to wait for futures to complete as part of the vm service api?